Invalid code generated: initialization discards ‘const’ qualifier from pointer target type

Source:
```haskell
test :: Def ('[ConstRef s ('Stored (Ptr 'Global ('Stored Uint8)))] ':-> ())
test = proc "test" $ \refptr -> body $ do
      ptr <- deref refptr
      withRef
        ptr
        (\ref -> do
           val <- deref ref
           store ref (val + 1))
        (pure ())
```

Generated:
```c
void test(const uint8_t **n_var0) {
    uint8_t *n_deref0 = *n_var0;
    if ((bool) (NULL != n_deref0)) {
        uint8_t n_deref1 = *n_deref0;       
        *n_deref0 = (uint8_t) (n_deref1 + (uint8_t) 1U);
    }
}
```

GCC error:
```
test.c: In function ‘test’:
test.c:8:25: error: initialization discards ‘const’ qualifier from pointer target type [-Werror=discarded-qualifiers]
     uint8_t *n_deref0 = *n_var0;
                         ^
```

Reason: `const a * *` (mutable pointer to mutable pointer to constant data) is wrong C encoding of type `ConstRef _ ('Stored (Ptr _ ('Stored a)))`, right encoding is `a * const *` (mutable pointer to constant pointer to mutable data).

--- a/Makefile
+++ b/Makefile
@@ -2,6 +2,8 @@ include stack.mk
 
 all: test
 
+STACK_CMD ?= stack
+
 IVORY_EX_TEST_DIR=test-dir-c-files
 
 PACKAGE= \
@@ -18,12 +20,22 @@ PACKAGE= \
 
 TEST_TARGETS=ivory-eval ivory-quickcheck
 
-.PHONY: test
-test: default
-	stack exec -- ivory-c-clang-test $(IVORY_EX_TEST_DIR)
+.PHONY: generic-test
+generic-test: default
+	$(STACK_CMD) build --test --no-run-tests --haddock --no-haddock-deps --pedantic
+	$(STACK_CMD) test $(TEST_TARGETS)
+	$(STACK_CMD) exec -- ivory-c-clang-test $(IVORY_EX_TEST_DIR)
 	cp ivory-examples/data/*.h $(IVORY_EX_TEST_DIR)/
-	cd $(IVORY_EX_TEST_DIR) && gcc -Wall -Wextra -I. -std=c99 -c *.c *.h -Wno-missing-field-initializers -Wno-unused-parameter -Wno-unused-variable -DIVORY_DEPLOY
-	stack test $(TEST_TARGETS)
+	cd $(IVORY_EX_TEST_DIR) &&			\
+		gcc	-Wall -Wextra -Werror		\
+			-Wno-missing-field-initializers	\
+			-Wno-unused-parameter		\
+			-Wno-unused-variable		\
+			-DIVORY_DEPLOY -I. -std=c99 -c	\
+			*.c *.h
+
+.PHONY: test
+test: generic-test
 
 .PHONY: sdist
 sdist:
@@ -34,14 +46,5 @@ veryclean:
 	stack clean
 	-rm -rf $(IVORY_EX_TEST_DIR)
 
-# Travis-ci specfic ############################################################
-
-TRAVIS_STACK ?= stack --no-terminal --system-ghc --skip-ghc-check
-
-travis-test:
-#	$(TRAVIS_STACK) build --test --no-run-tests --haddock --no-haddock-deps --pedantic
-	$(TRAVIS_STACK) build --test --no-run-tests --haddock --no-haddock-deps --ghc-options='-Werror'
-	$(TRAVIS_STACK) exec -- ivory-c-clang-test $(IVORY_EX_TEST_DIR)
-	cp ivory-examples/data/*.h $(IVORY_EX_TEST_DIR)/
-	cd $(IVORY_EX_TEST_DIR) && gcc -Wall -Wextra -I. -std=c99 -c *.c *.h -Wno-missing-field-initializers -Wno-unused-parameter -Wno-unused-variable -DIVORY_DEPLOY
-	$(TRAVIS_STACK) test $(TEST_TARGETS)
+travis-test: generic-test
+travis-test: STACK_CMD = stack --no-terminal --system-ghc --skip-ghc-check

--- a/ivory-backend-c/ivory-backend-c.cabal
+++ b/ivory-backend-c/ivory-backend-c.cabal
@@ -24,6 +24,7 @@ source-repository    this
 library
   exposed-modules:      Ivory.Compile.C,
                         Ivory.Compile.C.Gen,
+                        Ivory.Compile.C.Gen.Const,
                         Ivory.Compile.C.Modules,
                         Ivory.Compile.C.Prop,
                         Ivory.Compile.C.Types,

--- a/ivory-backend-c/runtime/ivory_asserts.h
+++ b/ivory-backend-c/runtime/ivory_asserts.h
@@ -54,11 +54,11 @@ extern void ivory_user_verbose_assert_hook(const char *asserttype,
 
 #else /* IVORY_TEST */
 
-#define REQUIRES(arg)
-#define ENSURES(arg)
-#define ASSUMES(arg)
-#define ASSERTS(arg)
-#define COMPILER_ASSERTS(arg)
+#define REQUIRES(arg)           (void)(arg)
+#define ENSURES(arg)            (void)(arg)
+#define ASSUMES(arg)            (void)(arg)
+#define ASSERTS(arg)            (void)(arg)
+#define COMPILER_ASSERTS(arg)   (void)(arg)
 
 #endif /* IVORY_TEST */
 

--- a/ivory-backend-c/src/Ivory/Compile/C/Gen.hs
+++ b/ivory-backend-c/src/Ivory/Compile/C/Gen.hs
@@ -1,5 +1,4 @@
 {-# LANGUAGE CPP            #-}
-{-# LANGUAGE PackageImports #-}
 {-# LANGUAGE QuasiQuotes    #-}
 
 -- | Ivory backend targeting language-c-quote.
@@ -7,13 +6,14 @@
 module Ivory.Compile.C.Gen where
 
 import           Language.C.Quote.GCC
-import qualified "language-c-quote" Language.C.Syntax                     as C
+import qualified Language.C.Syntax                     as C
 
 import qualified Ivory.Language.Array                  as I
 import qualified Ivory.Language.Proc                   as P
 import qualified Ivory.Language.Syntax                 as I
 import           Ivory.Language.Syntax.Concrete.Pretty
 
+import           Ivory.Compile.C.Gen.Const (makeTargetConstIf)
 import           Ivory.Compile.C.Prop
 import           Ivory.Compile.C.Types
 
@@ -146,7 +146,7 @@ toType = toType' False
 
 -- | C type conversion, with a special case for references and pointers.
 toType' :: Bool -> I.Type -> C.Type
-toType' b ty = case ty of
+toType' decay ty = case ty of
   I.TyVoid              -> [cty| void |]
   I.TyChar              -> [cty| char |]
   I.TyInt i             -> intSize i
@@ -162,18 +162,20 @@ toType' b ty = case ty of
     [cty| $ty:(toType retTy) (*)
           ($params:(map (toParam . toType) argTys)) |]
   I.TyOpaque            -> error "Opaque type is not implementable."
-  I.TyRef t             -> arrCase t
-  I.TyPtr t             -> arrCase t
-  I.TyConstRef t        -> [cty| const $ty:(arrCase t) |]
-  I.TyConstPtr t        -> [cty| const $ty:(arrCase t) |]
+  I.TyRef t             -> arrCase False  t
+  I.TyPtr t             -> arrCase False  t
+  I.TyConstRef t        -> arrCase True   t
+  I.TyConstPtr t        -> arrCase True   t
   where
-  arrCase t = case t of
-    I.TyArr len t'
-      -> if b then [cty| $ty:(toType t') * |]
-              else [cty| $ty:(toType t')[$uint:len] |]
-    I.TyCArray t'
-      -> [cty| $ty:(toType t') * |]
-    _ -> [cty| $ty:(toType t)  * |]
+  arrCase isTargetConst t =
+    makeTargetConstIf isTargetConst $
+    case t of
+      I.TyArr len t'
+        -> if decay then [cty| $ty:(toType t') * |]
+                    else [cty| $ty:(toType t')[$uint:len] |]
+      I.TyCArray t'
+        -> [cty| $ty:(toType t') * |]
+      _ -> [cty| $ty:(toType t)  * |]
 
 intSize :: I.IntSize -> C.Type
 intSize I.Int8  = [cty| typename int8_t  |]

--- a/ivory-backend-c/src/Ivory/Compile/C/Gen/Const.hs
+++ b/ivory-backend-c/src/Ivory/Compile/C/Gen/Const.hs
@@ -1,0 +1,56 @@
+module Ivory.Compile.C.Gen.Const (makeTargetConst, makeTargetConstIf) where
+
+import           Data.Loc (noLoc)
+import qualified Language.C.Syntax as C
+
+makeTargetConstIf :: Bool -> C.Type -> C.Type
+makeTargetConstIf c t = if c then makeTargetConst t else t
+
+makeTargetConst :: C.Type -> C.Type
+makeTargetConst = modifyTargetQuals (C.Tconst noLoc :)
+
+modifyTargetQuals :: ([C.TypeQual] -> [C.TypeQual]) -> C.Type -> C.Type
+modifyTargetQuals f typ@(C.Type _ decl _) = case decl of
+  C.Array _ _ decl1 _ -> go decl1
+  C.Ptr _ decl1 _     -> go decl1
+  _                   -> typ
+  where
+    go decl1 = case decl1 of
+      C.DeclRoot{} ->
+        -- `typ` is rank 1 pointer, modify base
+        (modifyTypeDeclSpec . modifyDeclSpecQuals) f typ
+      C.Ptr{} ->
+        -- `typ` is rank >1 pointer, modify second pointer level
+        (modifyTypeDecl . modifyNestedDecl . modifyPtrQuals) f typ
+      _ ->
+        -- nothing to do here
+        typ
+modifyTargetQuals _ typ@C.AntiType{} = internalError typ
+
+modifyDeclSpecQuals
+  :: ([C.TypeQual] -> [C.TypeQual]) -> C.DeclSpec -> C.DeclSpec
+modifyDeclSpecQuals f declSpec = case declSpec of
+  C.DeclSpec storage quals spec loc -> C.DeclSpec storage (f quals) spec loc
+  C.AntiDeclSpec{}                  -> internalError declSpec
+  C.AntiTypeDeclSpec{}              -> internalError declSpec
+
+modifyNestedDecl :: (C.Decl -> C.Decl) -> C.Decl -> C.Decl
+modifyNestedDecl f decl0 = case decl0 of
+  C.Array quals size decl loc -> C.Array quals size (f decl) loc
+  C.Ptr quals decl loc        -> C.Ptr quals (f decl) loc
+  _                           -> decl0
+
+modifyPtrQuals :: ([C.TypeQual] -> [C.TypeQual]) -> C.Decl -> C.Decl
+modifyPtrQuals f (C.Ptr quals decl loc)  = C.Ptr (f quals) decl loc
+modifyPtrQuals _ decl                    = decl
+
+modifyTypeDecl :: (C.Decl -> C.Decl) -> C.Type -> C.Type
+modifyTypeDecl f (C.Type declSpec decl loc) = C.Type declSpec (f decl) loc
+modifyTypeDecl _ typ@C.AntiType{}           = internalError typ
+
+modifyTypeDeclSpec :: (C.DeclSpec -> C.DeclSpec) -> C.Type -> C.Type
+modifyTypeDeclSpec f (C.Type declSpec decl loc) = C.Type (f declSpec) decl loc
+modifyTypeDeclSpec _ typ@C.AntiType{}           = internalError typ
+
+internalError :: Show a => a -> b
+internalError a = error $ "internal language-c-quote data leaked: " ++ show a

--- a/ivory-examples/data/some_header.h
+++ b/ivory-examples/data/some_header.h
@@ -1,0 +1,1 @@
+#include <stdio.h> // for `putchar`

--- a/ivory-examples/data/some_other_header.h
+++ b/ivory-examples/data/some_other_header.h
@@ -1,1 +1,1 @@
-#define SOME_CONST
+#define SOME_CONST 42

--- a/ivory-examples/examples/ConstPtrRef.hs
+++ b/ivory-examples/examples/ConstPtrRef.hs
@@ -1,0 +1,19 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE TypeOperators #-}
+
+module ConstPtrRef where
+
+import           Control.Applicative (pure)
+import           Ivory.Language
+
+test :: Def ('[ConstRef s ('Stored (Ptr 'Global ('Stored Uint8)))] ':-> ())
+test = proc "ConstPtrRef_test" $ \refptr -> body $ do
+  ptr <- deref refptr
+  withRef ptr
+    (\ref -> do
+      val <- deref ref
+      store ref $ val + 1)
+    (pure ())
+
+cmodule :: Module
+cmodule = package "ConstPtrRef" $ incl test

--- a/ivory-examples/examples/TestExamples.hs
+++ b/ivory-examples/examples/TestExamples.hs
@@ -9,6 +9,7 @@ import qualified Bits
 import qualified ClassHierarchy
 import qualified ConcreteFile
 import qualified Cond
+import qualified ConstPtrRef
 import qualified ConstRef
 import qualified Coroutine
 import qualified Extern
@@ -54,6 +55,7 @@ modules = [ AddrOfRegression.cmodule
           , ConcreteFile.concreteIvory
           , ConcreteFile.examplesfile
           , Cond.cmodule
+          , ConstPtrRef.cmodule
           , ConstRef.cmodule
           , Coroutine.cmodule
           , Extern.cmodule

--- a/ivory-examples/examples/TestExamples.hs
+++ b/ivory-examples/examples/TestExamples.hs
@@ -70,5 +70,4 @@ modules = [ AddrOfRegression.cmodule
           , SizeOf.cmodule
           , String.cmodule
           ]
-          ++ [stdlibStringModule]
           ++ stdlibModules

--- a/ivory-examples/examples/file.ivory
+++ b/ivory-examples/examples/file.ivory
@@ -59,7 +59,7 @@ struct Foo2
 r* struct ivory_string_FooStr foostr(uint8_t i, r* struct ivory_string_FooStr s) {
   -- Allocate a new string dynamic string. Fails if the string is too large.
   alloc s0{} = $stringInit("foo");
-  if (0>i) { foostr(i-1, s0); } else {}
+  if (0 < i) { foostr(i-1, s0); } else {}
   -- Store a new string into the dynamic string data structure.
   $string_lit_store("foos", s);
   return s;

--- a/ivory-examples/ivory-examples.cabal
+++ b/ivory-examples/ivory-examples.cabal
@@ -29,6 +29,7 @@ executable ivory-c-clang-test
                       , ClassHierarchy
                       , ConcreteFile
                       , Cond
+                      , ConstPtrRef
                       , ConstRef
                       , Coroutine
                       , Extern

